.doom.d/config.org
#+title: Doom Emacs Config
#+OPTIONS: toc:2

* General configuration

User information

#+begin_src elisp
(setq user-full-name "Nicholas Dunham"
      user-mail-address "nmdnhm@gmail.com")
#+end_src

* Utility functions
Expand file paths, regardless of how many elements you need to concat together and regardless of whether the last element is a file or directory.

#+begin_src elisp
(require 'cl-lib)
(defun nd/concat-path (&rest parts)
  (cl-reduce (lambda (a b) (expand-file-name b a)) parts))
#+end_src

Get a password, and only a password. Thanks to David Wilson of System Crafters for this code.

#+begin_src elisp
(defun nd/lookup-password (&rest keys)
  (let ((result (apply #' auth-source-search keys)))
    (if result
        (funcall (plist-get (car result) :secret))
      nil)))
#+end_src

* Aesthetics

** Fonts

I'm using two different versions of Fira Code, one proportional and one monospace.

#+begin_src elisp
(setq doom-font (font-spec :family "FiraCode Nerd Font" :size 12 :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 14)
      doom-unicode-font (font-spec :family "FiraCode Nerd Font" :size 12 :weight 'semi-light))
#+end_src

I also customize a few faces to make my agenda prettier.

#+begin_src elisp
(custom-set-faces!
  '(org-agenda-structure :height 1.30)
  '(org-super-agenda-header :inherit org-agenda-structure :height 0.9 :weight semi-bold))
#+end_src

** Themes

Customization of the Doom theme and modeline.

One thing that took me a while to figure out was what the option =doom-modeline-hud= does. I saw it in a few people's dotfiles, but with no explanation of what its purpose was. The docstring just says, "Whether to use hud instead of default bar. It's only respected in GUI." That's not very helpful if you don't know what "hud" or "default bar" mean in this context. I tried turning it on, but didn't immediately notice a difference. So in case anyone anyone stumbles across this trying to figure it out, here's my explanation.

By default, doom-modeline has a little colored bar on the far-left side of the modeline. Its primary purpose is to set the height of the modeline. You can change its width using the =doom-modeline-bar-width= variable, or (if you're using a GUI version of Emacs) you can switch it to a HUD, in which case it will show you approximately where you are in the document you're currently editing, sort of like a tiny scrollbar (except you can't actually use it to scroll through the file).

#+begin_src elisp
(setq doom-theme 'doom-nord
      doom-modeline-icon t
      doom-modeline-major-mode-icon t
      doom-modeline-major-mode-color-icon t
      doom-modeline-buffer-state-icon t
      doom-modeline-buffer-modification-icon t
      doom-modeline-enable-word-count t
      doom-modeline-continuous-word-count-modes '(markdown-mode gfm-mode org-mode)
      doom-modeline-buffer-encoding nil
      doom-modeline-checker-simple-format t
      doom-modeline-time nil
      doom-modeline-hud t)
#+end_src

** Window

Set a nice big window on startup.

#+begin_src elisp
(if (window-system) (set-frame-size (selected-frame) 150 80))
#+end_src

* Editor settings

** Line numbers

I use visual line numbers everywhere, and I want Evil mode to respect those line numbers and jump the appropriate number of visual lines, even when folding and soft line wrapping are in use. The =evil-respect-visual-line-mode= setting makes this happen, but it must be run before Evil mode is loaded, which can't be done from =config.el=. Loading it from =init.el= works. I'm leaving it here as well, even though it's redundant/ineffective here, for documentation purposes.

#+begin_src elisp
(setq-default display-line-numbers-type 'visual
              display-line-numbers-width 4
              line-move-visual t
              evil-respect-visual-line-mode t)
#+end_src

** Annoyances

By default, Doom Emacs replaces Vim's substitute command with Evil-snipe. I prefer to keep the default Vim behavior.

#+begin_src elisp
(remove-hook 'doom-first-input-hook #'evil-snipe-mode)
#+end_src

* Org mode

** Basic Org settings

#+begin_src elisp
(after! org
  (setq org-directory "~/org/"))
#+end_src

** Org beautification

Hide emphasis markers and generally make stuff pretty...

#+begin_src elisp
(after! org (setq org-hide-emphasis-markers t
                  org-pretty-entities t
                  org-pretty-entities-include-sub-superscripts t))
#+end_src

...but show emphasis markers when the cursor is on a word surrounded by them.

#+begin_src elisp
(add-hook! org-mode :append #'org-appear-mode)
#+end_src

*** Org fonts

Allow mixed fonts in Org buffers, for example to mix source and prose blocks in a document.

#+begin_src elisp
(use-package! mixed-pitch)
(add-hook! 'org-mode-hook #'mixed-pitch-mode)
#+end_src

Display text in a fixed pitch font where appropriate.

#+begin_src elisp
(after! org
  (custom-set-faces!
    '(org-block nil :foreground nil :inherit 'fixed-pitch)
    '(org-code nil   :inherit '(shadow fixed-pitch))
    '(org-table nil   :inherit '(shadow fixed-pitch))
    '(org-verbatim nil :inherit '(shadow fixed-pitch))
    '(org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
    '(org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))))
#+end_src

Make headings larger and bolder than paragraph text.

#+begin_src elisp
(after! org
  (custom-set-faces!
    '(org-document-title :height 1.5)
    '(org-level-1 :inherit outline-1 :weight extra-bold :height 1.4)
    '(org-level-2 :inherit outline-2 :weight bold :height 1.15)
    '(org-level-3 :inherit outline-3 :weight bold :height 1.12)
    '(org-level-4 :inherit outline-4 :weight bold :height 1.09)
    '(org-level-5 :inherit outline-5 :weight semi-bold :height 1.06)
    '(org-level-6 :inherit outline-6 :weight semi-bold :height 1.03)
    '(org-level-7 :inherit outline-7 :weight semi-bold)
    '(org-level-8 :inherit outline-8 :weight semi-bold)))
#+end_src

** GTD

*** Todo items

#+begin_src elisp
(after! org (setq org-log-done-with-time t
                  org-log-into-drawer "LOGBOOK"
                  org-log-refile 'time
                  org-log-repeat 'time
                  org-log-reschedule 'note
                  org-todo-keywords '((sequence "TODO(t)" "NEXT(n!)" "WAIT(w@/!)" "|" "DONE(d!)" "CNCL(c@)"))))

(after! org-fancy-priorities (setq org-fancy-priorities-list '("■" "■" "■")))
#+end_src

*** Agenda files

#+begin_src elisp
(after! org
  (setq org-agenda-files (append
                          '("inbox.org"
                            "emacs.org"
                            "work.org"
                            "lawsuit.org"
                            "personal.org")
                          (directory-files-recursively
                           (nd/concat-path org-directory "work-projects")
                           "\\.org$")
                          (directory-files-recursively
                           (nd/concat-path org-directory "education-projects")
                           "\\.org$"))
        org-refile-targets '((nil :maxlevel . 3)
                             (org-agenda-files :maxlevel . 3)
                             ("ideas.org" :maxlevel . 3))))
#+end_src

*** Org custom agenda

Trying this out for now.

#+begin_src elisp
(defun nd/agenda-auto-group-title-olp (item)
    (-when-let* ((title (cadar (org-collect-keywords '("title"))))
                 (ancestors (org-get-outline-path))
                 (top (nth 0 (reverse ancestors))))
      (concat title " → " top)))

(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq org-agenda-skip-scheduled-if-done t
        org-agenda-skip-deadline-if-done t
        org-agenda-include-deadlines t
        org-agenda-block-separator nil
        org-agenda-compact-blocks nil
        org-agenda-start-day nil ;; i.e. today
        org-agenda-start-on-weekday nil
        org-deadline-warning-days 0
        org-agenda-format-date "%A %e %B %Y")
  (setq org-agenda-custom-commands
        '(("c" "Super view"
           ((agenda "" ((org-agenda-overriding-header "Today\n")
                        (org-agenda-start-day nil)
                        (org-agenda-day-face-function (lambda (date) 'org-agenda-date))
                        (org-agenda-span 1)))
            (agenda "" ((org-agenda-overriding-header "\nNext three days\n")
                        (org-agenda-start-day "+1")
                        (org-agenda-span 3)))
            (agenda "" ((org-agenda-overriding-header "\nUpcoming deadlines\n")
                        (org-agenda-time-grid nil)
                        (org-agenda-start-day "+4")
                        (org-agenda-span 14)
                        (org-agenda-show-all-dates nil)
                        (org-agenda-entry-types '(:deadline))))
            (tags-todo "project"
                       ((org-agenda-overriding-header "\n\nProjects")
                         (org-super-agenda-groups
                          '((:discard (:deadline t :scheduled t))
                            (:auto-map nd/agenda-auto-group-title-olp)))))
           (tags-todo "education"
                       ((org-agenda-overriding-header "\n\nEducation")
                         (org-super-agenda-groups
                          '((:discard (:deadline t :scheduled t))
                            (:auto-map nd/agenda-auto-group-title-olp)))))
            (alltodo ""
                     ((org-agenda-overriding-header "\n\nOther tasks")
                      (org-super-agenda-groups
                       '((:discard (:deadline t :scheduled t :tag "project" :tag "education"))
                         (:name "Important and unscheduled"
                          :priority "A"
                          :order 1)
                         (:name "Miscellaneous work tasks"
                          :category "Work"
                          :order 2)
                         (:name "Household chores"
                          :tag "chore"
                          :order 3)
                         (:name "To refile"
                          :category "Inbox"
                          :order 4)))))))))
  :config
  (org-super-agenda-mode))
#+end_src

** Org annoyances

Avoid auto-completion in Org mode.

#+begin_src elisp
(defun nd/adjust-org-company-backends ()
  (remove-hook 'after-change-major-mode-hook '+company-init-backends-h)
  (setq-local company-backends nil))

(add-hook! org-mode (nd/adjust-org-company-backends))
#+end_src

I occasionally get the following error when opening Org Agenda: =org-element--cache: Unregistered buffer modifications detected. Resetting.= It doesn't seem to cause any problems displaying the agenda, and I'm not sure yet what's causing it---maybe manual edits to TODO entries rather than using Org's TODO management functions, but I haven't verified that, and it seems like the warning sometimes appears when I haven't made any changes at all.

There are a couple of ways I know of to get rid of the warning: add the warning type to =warning-suppress-types=, or disable caching.

For now, I'm going with suppressing the warning. I might try disabling caching later, just to compare the speed. I suspect that on any reasonably modern system, caching isn't noticeably speeding up Org.

#+begin_src elisp
(setq warning-suppress-types (append warning-suppress-types '((org-element-cache))))
#+end_src

=checkdoc= is a useful linter, but not when editing source blocks in Org files.

#+begin_src elisp
(defun nd/disable-flycheck-in-org-src-block ()
  (setq-local flycheck-disabled-checkers '(emacs-lisp-checkdoc)))

(add-hook 'org-src-mode-hook 'nd/disable-flycheck-in-org-src-block)
#+end_src

I want to log when I refile a task, but not when I refile it immediately from the capture buffer. This advice function sets =org-log-refile= to =nil= only when refiling from a capture buffer.

Weirdly, this also seems to solve a different problem I was having when refiling from the capture buffer: the buffer and the frame it was open in weren't closing when I refiled after capturing. In other words, if I captured to my inbox with =C-c C-c=, the buffer closed as expected, but if refiled the capture with =C-c C-w=, the capture buffer stayed open. That's no longer happening since I added this advice. I don't know why, but it's a nice bonus.

#+begin_src elisp
(define-advice org-capture-refile (:around (oldfunc &rest args) org-disable-log-refile)
  "Set `org-log-refile' to nil while capturing."
  (let ((org-log-refile nil))
    (apply oldfunc args)))
#+end_src

** Org-journal

*** General

#+begin_src elisp
(after! org-journal
  (setq org-journal-date-prefix "#+TITLE: "
        org-journal-time-prefix "* "
        org-journal-file-format "%Y-%m-%d.org"))
#+end_src

*** Capture

This function makes the Org-journal capture template work.

#+begin_src elisp
(defun nd/org-journal-find-location ()
  ;; Open today's journal, but specify a non-nil prefix argument in order to
  ;; inhibit inserting the heading; org-capture will insert the heading
  (org-journal-new-entry t)
  (unless (eq org-journal-file-type 'daily)
    (org-narrow-to-subtree))
  (goto-char (point-max)))
#+end_src

** Denote

I used to use Org-Roam, but I'm trying out Denote as a lightweight alternative. So far I like it.

#+begin_src elisp
(use-package! denote
  :after org
  :config
  (setq denote-directory (nd/concat-path org-directory "notes")
        denote-known-keywords '("work"
                                "french"
                                "ideas"
                                "python"
                                "ml&ai"
                                "clojure"
                                "elisp"
                                "geography")))
#+end_src

** Org-sticky-header

#+begin_src elisp
(use-package! org-sticky-header
  :after org
  :config
  (add-hook! 'org-mode-hook 'org-sticky-header-mode)
  (setq org-sticky-header-full-path 'full))
#+end_src

** Org-chef

Package for managing recipes in Org-mode.

#+begin_src elisp
(use-package! org-chef
  :after org)
#+end_src

** Capture templates

#+begin_src elisp
(after! org
  (setq org-template-dir (nd/concat-path doom-user-dir "templates")
        checkbox-template "- [ ] %?"
        org-capture-templates `(("t" "Task" entry
                                 (file "inbox.org")
                                 (file ,(nd/concat-path org-template-dir "capture-task.org"))
                                 :kill-buffer t)
                                ("m" "Meeting notes" entry
                                 (file "inbox.org")
                                 (file ,(nd/concat-path org-template-dir "capture-meeting.org"))
                                 :kill-buffer t)
                                ("n" "Note with Denote" plain
                                 (file denote-last-path)
                                 #'denote-org-capture
                                 :kill-buffer t)
                                ("s" "Shopping")
                                ("sg" "Groceries" checkitem
                                 (file+olp "shopping.org" "Groceries")
                                 ,checkbox-template
                                 :kill-buffer t)
                                ("so" "Online" checkitem
                                 (file+olp "shopping.org" "Online shopping")
                                 ,checkbox-template
                                 :kill-buffer t)
                                ("sl" "Liquor" checkitem
                                 (file+olp "shopping.org" "Liquor")
                                 ,checkbox-template
                                 :kill-buffer t)
                                ("sm" "Miscellaneous" checkitem
                                 (file+olp "shopping.org" "Miscellaneous")
                                 ,checkbox-template
                                 :kill-buffer t)
                                ("st" "Later" checkitem
                                 (file+olp "shopping.org" "Later")
                                 ,checkbox-template
                                 :kill-buffer t)
                                ("c" "Cookbook")
                                ("cc" "Capture recipe from web" entry
                                 (file+olp "cookbook.org" "Unfiled")
                                 "%(org-chef-get-recipe-from-url)"
                                 :kill-buffer t)
                                ("cm" "Enter recipe manually" entry
                                 (file+olp "cookbook.org" "Unfiled")
                                 (file ,(nd/concat-path org-template-dir "capture-recipe.org"))
                                 :kill-buffer t)
                                ("j" "Journal" plain
                                 (function nd/org-journal-find-location)
                                 "* %(format-time-string org-journal-time-format)%^{Title}\n%i%?"
                                 :kill-buffer))))
#+end_src

* Programming language support

** Lisp and its dialects

In the past I've tried Parinfer (great but took too much control from me and made some things unnecessarily difficult) and Lispy/Lispyville (required a lot of configuration, and annoyed me right off the bat by breaking the =jk -> Esc= mapping). Now I'm trying evil-cleverparens.

#+begin_src elisp
(add-hook!
 '(emacs-lisp-mode-hook lisp-mode-hook clojure-mode-hook)
 '(smartparens-strict-mode evil-cleverparens-mode))

(setq evil-move-beyond-eol t)
#+end_src

* Presenting and streaming
** Command log frame
I adapted this from David Wilson's dotfiles.

#+begin_src elisp
(use-package! command-log-mode)

(setq nd/command-frame nil)

(defun nd/toggle-command-frame ()
  (interactive)
  (if nd/command-frame
      (progn
        (posframe-delete-frame clm/command-log-buffer)
        (setq nd/command-frame nil))
    (progn
      (global-command-log-mode t)
      (with-current-buffer
          (setq clm/command-log-buffer
                (get-buffer-create " *command-log*"))
        (text-scale-set -1))
      (setq nd/command-frame
            (posframe-show clm/command-log-buffer
             :poshandler 'posframe-poshandler-frame-top-right-corner
             :width 45
             :height 5
             :min-width 45
             :min-height 5
             :internal-border-width 1
             :internal-border-color "#c792ea"
             :override-parameters '((parent-frame . nil)))))))

(map! :leader
      :desc "Command log frame"
      "t o" #'nd/toggle-command-frame)
#+end_src

** Org-tree-slide-mode

Great way to quickly build a presentation, but the default Doom configuration doesn't work at all for me. Fortunately, [[https://github.com/doomemacs/doomemacs/issues/4646#issuecomment-867763057][this GitHub comment]] has a solution.

#+begin_src elisp
(after! org-tree-slide
  (advice-remove 'org-tree-slide--display-tree-with-narrow
                 #'+org-present--hide-first-heading-maybe-a))

(defun nd/org-tree-slide-custom-settings ()
  (setq-local cursor-type nil
              display-line-numbers nil))

(add-hook! 'org-tree-slide-play-hook 'nd/org-tree-slide-custom-settings)

#+end_src

* Other fun stuff

** GPT-3 in Emacs

OpenAI's GPT-3 API in Emacs!

The =nd/openai-api-get-engines= helper function gets a list of available engines and prints it to the =*Messages*= buffer. It's good to check once in a while to see if OpenAI has released any new ones.

The =nd/openai-api-complete-region= function is a wrapper around the openai-api plugin's built-in function of the same name. Its purpose is to set the =openai-api-secret-key= variable on run, if it's not already set. Previously I was trying to set that variable on startup, which was causing problems.

To use, highlight a prompt and run the =nd/openai-api-complete-region= function (I have this mapped to =SPC-i-g=). The response will be inserted on the next line.

So far, using these settings, this is costing about a penny per run on average.

#+begin_src elisp
(defun nd/openai-api-get-engines (&optional no-cache)
  "Write a list of available engines to the messages buffer"
  (interactive)
  (message "%s" (openai-api-get-engines no-cache)))

(defun nd/openai-api-complete-region ()
  (interactive)
  (unless openai-api-secret-key
    (setq openai-api-secret-key (nd/lookup-password :host "openai")))
  (openai-api-complete-region))

(use-package! openai-api
  :config
  (setq openai-api-engine "text-davinci-002"
        openai-api-completion-params '((max_tokens . 1024)
                                       (temperature . 0.7)
                                       (frequency_penalty . 0)
                                       (presence_penalty . 0)
                                       (n . 1))))

(map! :leader
      :desc "Response from GPT-3"
      "i g" #'nd/openai-api-complete-region)
#+end_src

** Atomic Chrome

Allows me to edit text areas on the web using Emacs. Relies on the [[https://chrome.google.com/webstore/detail/ghosttext/godiecgffnchndlihlpaajjcplehddca][GhostText]] Chrome extension.

#+begin_src elisp
(use-package! atomic-chrome
  :config (atomic-chrome-start-server))
#+end_src

** Csound-mode

#+begin_src elisp
(use-package! csound-mode)
#+end_src


.doom.d/config.el
(setq user-full-name "Nicholas Dunham"
      user-mail-address "nmdnhm@gmail.com")

(require 'cl-lib)
(defun nd/concat-path (&rest parts)
  (cl-reduce (lambda (a b) (expand-file-name b a)) parts))

(setq doom-font (font-spec :family "FiraCode Nerd Font" :size 12 :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 14)
      doom-unicode-font (font-spec :family "FiraCode Nerd Font" :size 12 :weight 'semi-light))

(setq doom-theme 'doom-nord
      doom-modeline-icon t
      doom-modeline-major-mode-icon t
      doom-modeline-major-mode-color-icon t
      doom-modeline-buffer-state-icon t
      doom-modeline-buffer-modification-icon t
      doom-modeline-enable-word-count t
      doom-modeline-continuous-word-count-modes '(markdown-mode gfm-mode org-mode)
      doom-modeline-buffer-encoding nil
      doom-modeline-checker-simple-format t
      doom-modeline-time nil)

(if (window-system) (set-frame-size (selected-frame) 150 80))

(global-display-line-numbers-mode)
(global-visual-line-mode)

(setq display-line-numbers-type 'visual
      line-move-visual t
      evil-respect-visual-line-mode t)

(add-hook 'vterm-mode-hook 'display-line-numbers-mode)
(add-hook 'term-mode-hook 'display-line-numbers-mode)

(remove-hook 'doom-first-input-hook #'evil-snipe-mode)

(after! org
  (setq org-directory "~/Projects/org/"))

(after! org (setq org-hide-emphasis-markers t
                  org-pretty-entities t
                  org-pretty-entities-include-sub-superscripts t))

(add-hook! org-mode :append #'org-appear-mode)

(use-package! mixed-pitch)
(add-hook! 'org-mode-hook #'mixed-pitch-mode)

(after! org
  (custom-set-faces!
    '(org-block nil :foreground nil :inherit 'fixed-pitch)
    '(org-code nil   :inherit '(shadow fixed-pitch))
    '(org-table nil   :inherit '(shadow fixed-pitch))
    '(org-verbatim nil :inherit '(shadow fixed-pitch))
    '(org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
    '(org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))))

(after! org
  (custom-set-faces!
    '(org-document-title :height 1.5)
    '(org-level-1 :inherit outline-1 :weight extra-bold :height 1.4)
    '(org-level-2 :inherit outline-2 :weight bold :height 1.15)
    '(org-level-3 :inherit outline-3 :weight bold :height 1.12)
    '(org-level-4 :inherit outline-4 :weight bold :height 1.09)
    '(org-level-5 :inherit outline-5 :weight semi-bold :height 1.06)
    '(org-level-6 :inherit outline-6 :weight semi-bold :height 1.03)
    '(org-level-7 :inherit outline-7 :weight semi-bold)
    '(org-level-8 :inherit outline-8 :weight semi-bold)))

(after! org (setq org-log-done-with-time t
                  org-log-into-drawer "LOGBOOK"
                  org-log-refile 'time
                  org-log-repeat 'time
                  org-log-reschedule 'note
                  org-todo-keywords '((sequence "TODO(t)" "NEXT(n!)" "WAIT(w@/!)" "|" "DONE(d!)" "CNCL(c@)"))))

(after! org-fancy-priorities (setq org-fancy-priorities-list '("■" "■" "■")))

(after! org
  (setq org-agenda-files '("inbox.org"
                           "emacs.org"
                           "work.org"
                           "personal.org")
        org-refile-targets '((nil :maxlevel . 3)
                             (org-agenda-files :maxlevel . 3)
                             ("ideas.org" :maxlevel . 3))))

(setq org-agenda-custom-commands
      '(("A" "Daily agenda and top priority tasks"
         ((tags-todo "+PRIORITY=A"
                     ((org-agenda-skip-function '(org-agenda-skip-entry-if 'scheduled 'deadline))
                      (org-agenda-block-separator nil)
                      (org-agenda-overriding-header "Important tasks without a date\n")))
          (todo "WAIT"
                ((org-agenda-block-separator nil)
                 (org-agenda-skip-function '(org-agenda-skip-if nil '(timestamp)))
                 (org-agenda-overriding-header "\nUndated tasks on hold\n")))
          (agenda ""
                 ((org-agenda-block-separator nil)
                  (org-agenda-start-day "+0d")
                  (org-agenda-span 1)
                  (org-deadline-warning-days 0)
                  (org-scheduled-past-days 0)
                  (org-agenda-day-face-function (lambda (date) 'org-agenda-date))
                  (org-agenda-format-date "%A %-e %B %Y")
                  (org-agenda-overriding-header "\nDaily agenda\n")))
          (agenda ""
                  ((org-agenda-start-on-weekday nil)
                   (org-agenda-start-day "+1d")
                   (org-agenda-span 3)
                   (org-deadline-warning-days 0)
                   (org-agenda-block-separator nil)
                   (org-agenda-skip-function '(org-agenda-skip-entry-if 'todo 'done))
                   (org-agenda-overriding-header "\nNext three days\n")))
          (agenda ""
                  ((org-agenda-time-grid nil)
                   (org-agenda-start-on-weekday nil)
                   (org-agenda-start-day "+3d")
                   (org-agenda-span 14)
                   (org-agenda-show-all-dates nil)
                   (org-deadline-warning-days 0)
                   (org-agenda-block-separator nil)
                   (org-agenda-entry-types '(:deadline))
                   (org-agenda-skip-function '(org-agenda-skip-entry-if 'todo 'done))
                   (org-agenda-overriding-header "\nUpcoming deadlines (+14d)\n")))))))

(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq org-agenda-skip-scheduled-if-done t
        org-agenda-skip-deadline-if-done t
        org-agenda-include-deadlines t
        org-agenda-block-separator nil
        org-agenda-compact-blocks nil
        org-agenda-start-day nil ;; i.e. today
        org-agenda-start-on-weekday nil
        org-deadline-warning-days 0
        org-agenda-format-date "%A %e %B %Y")
  (setq org-agenda-custom-commands
        '(("c" "Super view"
           ((agenda "" ((org-agenda-overriding-header "Today\n")
                        (org-agenda-start-day nil)
                        (org-agenda-day-face-function (lambda (date) 'org-agenda-date))
                        (org-agenda-span 1)))
            (agenda "" ((org-agenda-overriding-header "\nNext three days\n")
                        (org-agenda-start-day "+1")
                        (org-agenda-span 3)))
            (agenda "" ((org-agenda-overriding-header "\nUpcoming deadlines\n")
                        (org-agenda-time-grid nil)
                        (org-agenda-start-day "+4")
                        (org-agenda-span 14)
                        (org-agenda-show-all-dates nil)
                        (org-agenda-entry-types '(:deadline))))
            (alltodo "" ((org-agenda-overriding-header "\nUndated tasks\n")
                         (org-super-agenda-groups
                          '((:discard (:deadline t :scheduled t))
                            (:name "\nImportant\n"
                             :priority "A")
                            (:auto-parent t
                             :name "\nActive work projects\n"
                             :and (:category "Work" :tag "Project"))
                            (:name "\nTo refile\n"
                             :category "Inbox")
                            (:name "\nOn hold\n"
                             :todo "WAIT")))))))))
  :init
  (org-super-agenda-mode))

(defun nd/adjust-org-company-backends ()
  (remove-hook 'after-change-major-mode-hook '+company-init-backends-h)
  (setq-local company-backends nil))

(add-hook! org-mode (nd/adjust-org-company-backends))

(after! org-journal
  (setq org-journal-date-prefix "#+TITLE: "
        org-journal-time-prefix "* "
        org-journal-file-format "%Y-%m-%d.org"))

(defun nd/org-journal-find-location ()
  ;; Open today's journal, but specify a non-nil prefix argument in order to
  ;; inhibit inserting the heading; org-capture will insert the heading
  (org-journal-new-entry t)
  (unless (eq org-journal-file-type 'daily)
    (org-narrow-to-subtree))
  (goto-char (point-max)))

(use-package! denote
  :after org
  :config
  (setq denote-directory (nd/concat-path org-directory "notes")
        denote-known-keywords '("work"
                                "french"
                                "ideas"
                                "python"
                                "ml&ai"
                                "clojure"
                                "elisp"
                                "geography")))

(after! org
  (setq org-capture-templates '(("i" "Inbox" entry
                                 (file "inbox.org")
                                 "* TODO %i%?\n:PROPERTIES:\n:CREATED: %U\n:END:\n:LOGBOOK:\n:END:"
                                 :kill-buffer t)
                                ("s" "Shopping")
                                ("sg" "Groceries" checkitem
                                 (file+olp "shopping.org" "Groceries")
                                 "- [ ] %^{Item}"
                                 :immediate-finish t)
                                ("so" "Online" checkitem
                                 (file+olp "shopping.org" "Online shopping")
                                 "- [ ] %^{Item}"
                                 :immediate-finish t)
                                ("sl" "Liquor" checkitem
                                 (file+olp "shopping.org" "Liquor")
                                 "- [ ] %^{Item}"
                                 :immediate-finish t)
                                ("sm" "Miscellaneous" checkitem
                                 (file+olp "shopping.org" "Miscellaneous")
                                 "- [ ] %^{Item}"
                                 :immediate-finish t)
                                ("st" "Later" checkitem
                                 (file+olp "shopping.org" "Later")
                                 "- [ ] %^{Item}"
                                 :immediate-finish t)
                                ("n" "Note with Denote" plain
                                 (file denote-last-path)
                                 #'denote-org-capture
                                 :no-save t
                                 :immediate-finish nil
                                 :kill-buffer t
                                 :jump-to-captured t)
                                ("j" "Journal" plain
                                 (function nd/org-journal-find-location)
                                 "* %(format-time-string org-journal-time-format)%^{Title}\n%i%?"
                                 :jump-to-captured t
                                 :immediate-finish t))))

(use-package! command-log-mode)

(setq nd/command-frame nil)

(defun nd/toggle-command-frame ()
  (interactive)
  (if nd/command-frame
      (progn
        (posframe-delete-frame clm/command-log-buffer)
        (setq nd/command-frame nil))
    (progn
      (global-command-log-mode t)
      (with-current-buffer
          (setq clm/command-log-buffer
                (get-buffer-create " *command-log*"))
        (text-scale-set -1))
      (setq nd/command-frame
            (posframe-show clm/command-log-buffer
             :poshandler 'posframe-poshandler-frame-top-right-corner
             :width 45
             :height 5
             :min-width 45
             :min-height 5
             :internal-border-width 1
             :internal-border-color "#c792ea"
             :override-parameters '((parent-frame . nil)))))))

(map! :leader
      :desc "Command log frame"
      "t o" #'nd/toggle-command-frame)
